src/transports/ssh_backend.h

```diff
@@ -0,0 +1,477 @@
+	if (rc != SSH_OK) {
+
+	}
+
+	rc = ssh_userauth_publickey(session, c->username, NULL);
+	if (rc != SSH_OK) {
+
+	}
+
+	return rc;
+#endif
+}
+
+static inline int _git_ssh_authenticate_session(
+	git_ssh_session session,
+	git_cred *cred)
+{
+	int rc;
+
+	do {
+		git_error_clear();
+		switch (cred->credtype) {
+			case GIT_CREDTYPE_USERPASS_PLAINTEXT: {
+				git_cred_userpass_plaintext *c = (git_cred_userpass_plaintext *)cred;
+#ifdef GIT_LIBSSH2
+				rc = libssh2_userauth_password(session, c->username, c->password);
+#elif defined(GIT_LIBSSH)
+				rc = ssh_userauth_password(session, c->username, c->password);
+#endif
+				break;
+			}
+			case GIT_CREDTYPE_SSH_KEY: {
+				git_cred_ssh_key *c = (git_cred_ssh_key *)cred;
+
+				if (c->privatekey) {
+#ifdef GIT_LIBSSH2
+					rc = libssh2_userauth_publickey_fromfile(
+						session, c->username, c->publickey,
+						c->privatekey, c->passphrase);
+#elif defined(GIT_LIBSSH)
+					ssh_key privkey;
+
+					rc = ssh_pki_import_privkey_file(c->privatekey, c->passphrase, NULL, NULL, &privkey);
+					if (rc != SSH_OK)
+						break;
+
+					rc = ssh_userauth_publickey(session, c->username, privkey);
+
+					ssh_key_free(privkey);
+#endif
+				} else
+					rc = git__ssh_agent_auth(session, c);
+
+				break;
+			}
+			case GIT_CREDTYPE_SSH_CUSTOM: {
+				git_cred_ssh_custom *c = (git_cred_ssh_custom *)cred;
+
+#ifdef GIT_LIBSSH2
+				rc = libssh2_userauth_publickey(
+					session, c->username, (const unsigned char *)c->publickey,
+					c->publickey_len, c->sign_callback, &c->payload);
+#elif defined(GIT_LIBSSH)
+				/* AFAICS there's no sign callback stuff */
+				GIT_UNUSED(c);
+
+				rc = -1;
+#endif
+				break;
+			}
+			case GIT_CREDTYPE_SSH_INTERACTIVE: {
+				git_cred_ssh_interactive *c = (git_cred_ssh_interactive *)cred;
+#ifdef GIT_LIBSSH2
+				void **abstract = libssh2_session_abstract(session);
+
+				/* ideally, we should be able to set this by calling
+				 * libssh2_session_init_ex() instead of libssh2_session_init().
+				 * libssh2's API is inconsistent here i.e. libssh2_userauth_publickey()
+				 * allows you to pass the `abstract` as part of the call, whereas
+				 * libssh2_userauth_keyboard_interactive() does not!
+				 *
+				 * The only way to set the `abstract` pointer is by calling
+				 * libssh2_session_abstract(), which will replace the existing
+				 * pointer as is done below. This is safe for now (at time of writing),
+				 * but may not be valid in future.
+				 */
+				*abstract = c->payload;
+
+				rc = libssh2_userauth_keyboard_interactive(
+					session, c->username, c->prompt_callback);
+#elif defined(GIT_LIBSSH)
+
+				rc = ssh_userauth_kbdint(session, c->username, NULL);
+				if (rc == SSH_AUTH_INFO) {
+					const char *name = ssh_userauth_kbdint_getname(session);
+					const char *instr = ssh_userauth_kbdint_getinstruction(session);
+					int num_prompts = ssh_userauth_kbdint_getnprompts(session);
+					size_t i;
+					ssh_userauth_kbdint_getprompt(session, i, NULL);
```

Comment:
This is the API endpoint I'm having issues with — the public callback with the libssh2 types — with the libssh API.

 I'm not sure how to best proceed forward from here :
- break the API and make a more generic callback, with no exposed types from the backends (the new `git_ssh_` types).
- provide a libssh-specific version of it, passing only the session. This is contrary to the guideline that the API should not depend on preprocessing, and would cause Docurium to miss it (as an example).
- make the SSH session accessible, so we don't have to wrap so closely libssh2, and deprecate anything that's not really needed.

Any suggestions ?